ci/L0_additional_outputs_vllm/additional_outputs_test.py

```diff
@@ -0,0 +1,198 @@
+            self.assertNotEqual(cumulative_logprob, prev_cumulative_logprob)
+            prev_cumulative_logprob = cumulative_logprob
+
+    def _assert_num_token_ids(self, output_num_token_ids):
+        for response in self._responses:
+            result, error = response["result"], response["error"]
+            self.assertIsNone(error)
+            num_token_ids_np = result.as_numpy(name="num_token_ids")
+            if output_num_token_ids is None or output_num_token_ids == False:
+                self.assertIsNone(num_token_ids_np)
+                continue
+            num_token_ids = num_token_ids_np[0].astype(int)
+            # TODO: vLLM may return token ids identical to the previous one when
+            #       streaming, for example:
+            #
+            #       prev: None
+            #       curr: text=' the', token_ids=array('l', [5])
+            #
+            #       prev: text=' the', token_ids=array('l', [5, 1385])
+            #       curr: text=' the term', token_ids=array('l', [5, 1385])
+            #
+            #       prev: text=' the term', token_ids=array('l', [5, 1385, 44])
+            #       curr: text=' the term', token_ids=array('l', [5, 1385, 44])
+            #
+            #       prev: text=' the term', token_ids=array('l', [5, 1385, 44, 48])
+            #       curr: text=' the term “', token_ids=array('l', [5, 1385, 44, 48])
+            #
+            #       If this is no longer the case in a future release, change the assert
+            #       to assertGreater().
+            self.assertGreaterEqual(num_token_ids, 0)
+
+    def _assert_additional_outputs_valid(
+        self,
+        stream,
+        output_finish_reason,
+        output_cumulative_logprob,
+        output_num_token_ids,
+    ):
+        inputs = self._get_inputs(
+            self._prompt,
+            stream=stream,
+            sampling_parameters=self._sampling_parameters,
+            output_finish_reason=output_finish_reason,
+            output_cumulative_logprob=output_cumulative_logprob,
+            output_num_token_ids=output_num_token_ids,
+        )
+        self._llm_infer(inputs)
+        self._assert_text_output_valid()
+        self._assert_finish_reason(output_finish_reason)
+        self._assert_cumulative_logprob(output_cumulative_logprob)
+        self._assert_num_token_ids(output_num_token_ids)
+
+    def test_additional_outputs(self):
+        for stream in [True, False]:
+            choices = [None, False, True]
+            for output_finish_reason in choices:
+                for output_cumulative_logprob in choices:
+                    for output_num_token_ids in choices:
+                        self._assert_additional_outputs_valid(
+                            stream,
+                            output_finish_reason,
+                            output_cumulative_logprob,
+                            output_num_token_ids,
+                        )
```

Comment:
FYI if your goal is to test all of the combinations, `pytest` has some nice features for this, such as `pytest.parametrize`:

![image](https://github.com/user-attachments/assets/c4ebfce1-b2ff-46fd-9f06-8773076a2a2e)

Reply:
```python
import pytest

class TestOptionalOutputs:
    @pytest.mark.parametrize("stream", [True, False])
    @pytest.mark.parametrize("output_finish_reason", [None, True, False])
    @pytest.mark.parametrize("output_cumulative_logprob", [None, True, False])
    @pytest.mark.parametrize("output_num_token_ids", [None, True, False])
    def test_additional_outputs(self, stream, output_finish_reason, output_cumulative_logprob, output_num_token_ids):
        print(
            stream,
            output_finish_reason,
            output_cumulative_logprob,
            output_num_token_ids,
        )
```

Reply:
The outputs from pytest is cleaner, let us [Switch to pytest](https://github.com/triton-inference-server/vllm_backend/pull/70/commits/44edd6e3cc1b8393d13a5b9256697e6609386259)

Reply:
```
+ python3 -m pytest -s -v additional_outputs_test.py
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.1.1, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /opt/tritonserver/vllm_backend
configfile: pyproject.toml
plugins: anyio-4.6.2.post1
collecting ... 
additional_outputs_test.py::TestAdditionalOutputs::test_additional_outputs[None-None-None-True] PASSED
...
additional_outputs_test.py::TestAdditionalOutputs::test_additional_outputs[False-False-False-False] PASSED
============================== 54 passed in 6.30s ==============================
```